### PR TITLE
io_uring: don't remove test tmp directory manually

### LINKED ISSUE #20764: io_uring tests run locally fail on tear-down
*Title*: io_uring tests run locally fail on tear-down

*Description*:
When running the io_uring tests locally (`bazel test //test/common/io:io_uring_impl_test`) we get the following error:
```
==========] Running 9 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 4 tests from IoUringImplTest
[ RUN      ] IoUringImplTest.Instantiate
[2022-04-10 04:38:54.586][12][critical][assert] [test/test_common/environment.cc:156] assert failure: ::rmdir(path.c_str()) == 0. Details: failed to remove path: /home/adip/.cache/bazel/_bazel_adip/1a87578f218430491883bb8ba9a0acc3/sandbox/linux-sandbox/538/execroot/envoy/_tmp/27458d54c9da232414e69eaa32c5d01e (rmdir failed)
[2022-04-10 04:38:54.586][12][critical][backtrace] [./source/server/backtrace.h:104] Caught Aborted, suspect faulting address 0x3eb0000000c
[2022-04-10 04:38:54.586][12][critical][backtrace] [./source/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2022-04-10 04:38:54.586][12][critical][backtrace] [./source/server/backtrace.h:92] Envoy version: 0/1.22.0-dev/test/DEBUG/BoringSSL
[2022-04-10 04:38:54.596][12][critical][backtrace] [./source/server/backtrace.h:96] #0: Envoy::SignalAction::sigHandler() [0x32adcb8]
[2022-04-10 04:38:54.596][12][critical][backtrace] [./source/server/backtrace.h:96] #1: __restore_rt [0x7fbc47f881f0]
[2022-04-10 04:38:54.605][12][critical][backtrace] [./source/server/backtrace.h:96] #2: Envoy::Io::(anonymous namespace)::IoUringImplTest::TearDown() [0x1656760]
[2022-04-10 04:38:54.614][12][critical][backtrace] [./source/server/backtrace.h:96] #3: testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x3f798ab]
[2022-04-10 04:38:54.623][12][critical][backtrace] [./source/server/backtrace.h:96] #4: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x3f69abd]
[2022-04-10 04:38:54.633][12][critical][backtrace] [./source/server/backtrace.h:96] #5: testing::Test::Run() [0x3f519e0]
[2022-04-10 04:38:54.642][12][critical][backtrace] [./source/server/backtrace.h:96] #6: testing::TestInfo::Run() [0x3f525aa]
[2022-04-10 04:38:54.651][12][critical][backtrace] [./source/server/backtrace.h:96] #7: testing::TestSuite::Run() [0x3f52e0b]
[2022-04-10 04:38:54.660][12][critical][backtrace] [./source/server/backtrace.h:96] #8: testing::internal::UnitTestImpl::RunAllTests() [0x3f61c4b]
[2022-04-10 04:38:54.670][12][critical][backtrace] [./source/server/backtrace.h:96] #9: testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x3f7c28b]
[2022-04-10 04:38:54.679][12][critical][backtrace] [./source/server/backtrace.h:96] #10: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x3f6bef3]
[2022-04-10 04:38:54.688][12][critical][backtrace] [./source/server/backtrace.h:96] #11: testing::UnitTest::Run() [0x3f6177e]
[2022-04-10 04:38:54.697][12][critical][backtrace] [./source/server/backtrace.h:96] #12: RUN_ALL_TESTS() [0x2f2f3f1]
[2022-04-10 04:38:54.707][12][critical][backtrace] [./source/server/backtrace.h:96] #13: Envoy::TestRunner::RunTests() [0x2f2eac0]
[2022-04-10 04:38:54.716][12][critical][backtrace] [./source/server/backtrace.h:96] #14: main [0x2f2b456]
[2022-04-10 04:38:54.716][12][critical][backtrace] [./source/server/backtrace.h:96] #15: __libc_start_main [0x7fbc47db0565]
```

This seems to be an issue with the test fixture when removing the temporary directory.

This isn't reproduced when executing the test using RBE, and the tests pass.

### PR DESCRIPTION
**Pull request by rojkov**
Commit Message: io_uring: don't remove test tmp directory manually
Additional Description:
Temp directory is created once for a test binary and removing it simultaneously from multiple test cases fails.
Don't remove it manually, let bazel delete it.

Risk Level: low
Testing: run unit tests
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Fixes #20764

### CODE REVIEW

**Comment by rojkov** [comment]
@adisuissa could you please check if this fix helps with the issue?

**Comment by adisuissa** [comment]
After applying the patch I got the following error:
```
[==========] Running 9 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 4 tests from IoUringImplTest
[ RUN      ] IoUringImplTest.Instantiate
[       OK ] IoUringImplTest.Instantiate (38 ms)
[ RUN      ] IoUringImplTest.RegisterEventfd
[       OK ] IoUringImplTest.RegisterEventfd (684 ms)
[ RUN      ] IoUringImplTest.PrepareReadvAllDataFitsOneChunk
[2022-04-13 03:57:17.138][12][critical][assert] [test/test_common/environment.cc:156] assert failure: ::rmdir(path.c_str()) == 0. Details: failed to remove path: /usr/local/google/home/adip/.cache/bazel/_bazel_adip/27316da064ea9a561ed452db3a62fafe/sandbox/linux-sandbox/436/execroot/envoy/_tmp/27458d54c9da232414e69eaa32c5d01e (rmdir failed)
[2022-04-13 03:57:17.138][12][critical][backtrace] [./source/server/backtrace.h:104] Caught Aborted, suspect faulting address 0xb9bae0000000c
[2022-04-13 03:57:17.138][12][critical][backtrace] [./source/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2022-04-13 03:57:17.138][12][critical][backtrace] [./source/server/backtrace.h:92] Envoy version: 0/1.22.0-dev/test/DEBUG/BoringSSL
[2022-04-13 03:57:17.163][12][critical][backtrace] [./source/server/backtrace.h:96] #0: Envoy::SignalAction::sigHandler() [0x32c8f18]
[2022-04-13 03:57:17.163][12][critical][backtrace] [./source/server/backtrace.h:96] #1: __restore_rt [0x7f3c9603c200]
[2022-04-13 03:57:17.189][12][critical][backtrace] [./source/server/backtrace.h:96] #2: Envoy::Io::(anonymous namespace)::IoUringImplTest_PrepareReadvAllDataFitsOneChunk_Test::TestBody() [0x167351f]
[2022-04-13 03:57:17.214][12][critical][backtrace] [./source/server/backtrace.h:96] #3: testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x3fb0beb]
[2022-04-13 03:57:17.240][12][critical][backtrace] [./source/server/backtrace.h:96] #4: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x3fa0e5d]
[2022-04-13 03:57:17.266][12][critical][backtrace] [./source/server/backtrace.h:96] #5: testing::Test::Run() [0x3f88d43]
[2022-04-13 03:57:17.292][12][critical][backtrace] [./source/server/backtrace.h:96] #6: testing::TestInfo::Run() [0x3f8994a]
[2022-04-13 03:57:17.317][12][critical][backtrace] [./source/server/backtrace.h:96] #7: testing::TestSuite::Run() [0x3f8a1ab]
[2022-04-13 03:57:17.342][12][critical][backtrace] [./source/server/backtrace.h:96] #8: testing::internal::UnitTestImpl::RunAllTests() [0x3f98feb]
[2022-04-13 03:57:17.368][12][critical][backtrace] [./source/server/backtrace.h:96] #9: testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x3fb35cb]
[2022-04-13 03:57:17.393][12][critical][backtrace] [./source/server/backtrace.h:96] #10: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x3fa3293]
[2022-04-13 03:57:17.421][12][critical][backtrace] [./source/server/backtrace.h:96] #11: testing::UnitTest::Run() [0x3f98b1e]
[2022-04-13 03:57:17.447][12][critical][backtrace] [./source/server/backtrace.h:96] #12: RUN_ALL_TESTS() [0x2f48801]
[2022-04-13 03:57:17.473][12][critical][backtrace] [./source/server/backtrace.h:96] #13: Envoy::TestRunner::RunTests() [0x2f47ed0]
[2022-04-13 03:57:17.500][12][critical][backtrace] [./source/server/backtrace.h:96] #14: main [0x2f44866]
[2022-04-13 03:57:17.500][12][critical][backtrace] [./source/server/backtrace.h:96] #15: __libc_start_main [0x7f3c95e747fd]
```

So it doesn't fail on the first test, but fails when `removePath` is called.
Can the path be non-empty?

**Comment by rojkov** [comment]
Weird... the path is non-empty of course, but `removePath` is a recursive function which deletes the content first, then the directory itself.

I tried to run the test inside a docker container (based on the CI image) and outside, via `bazel run` and directly and still couldn't reproduce it. What's your environment?

**Comment by adisuissa** [comment]
It's a company standard Linux-based dev-env, but don't know what specific details I can provide to debug this issue.
Note that it only happens in a local build, and not when RBE is used.
It's also hard to debug, because the bug only happens when running `bazel test //test/common/io:io_uring_impl_test`.
If I build it first (`bazel build`) and then execute the test-binary, it passes.

I think that Matt also observed the same problem, and he is using a different dev-env.

**Comment by rojkov** [comment]
Ok, thanks! I reproduced it with `bazel test` too outside of docker. `errno` is `EBUSY`. Looks like the temp directory is created once for a test binary and removing it simultaneously from multiple test cases fails. The other tests using `TestEnvironment::writeStringToFileForTest()` seem not to bother removing the dir. So, I did the same.

File: test/common/io/io_uring_impl_test.cc [hunk 1]
```diff
@@ -29,7 +29,7 @@ class IoUringBaseTest : public ::testing::Test {
 };
 
 class IoUringImplParamTest
-    : public IoUringBaseTest,
+    : public IoUringImplTest,
```
Comment (adisuissa) [inline test/common/io/io_uring_impl_test.cc:32]: nit: merge  `IoUringImplParamTest` and `IoUringImplTest` (unless you plan to have more extensions.

**Review by adisuissa — COMMENTED** [review]
Thanks for tracking this down!

**Review by adisuissa — APPROVED** [review]
Thanks for fixing this!

**Pull request merged**